cpp/src/experimental/random_walks.cuh

```diff
@@ -0,0 +1,854 @@
+ * @param d_v_start Device set of starting vertex indices for the RW.
+ * number(paths) == d_v_start.size().
+ * @param max_depth maximum length of RWs.
+ * @return std::tuple<device_vec_t<vertex_t>, device_vec_t<weight_t>,
+ * device_vec_t<index_t>> Triplet of coalesced RW paths, with corresponding edge weights for
+ * each, and coresponding path sizes. This is meant to minimize the number of DF's to be passed to
+ * the Python layer.
+ */
+template <typename graph_t,
+          typename random_engine_t =
+            rrandom_gen_t<typename graph_t::vertex_type, typename graph_t::edge_type>,
+          typename index_t = typename graph_t::edge_type>
+std::enable_if_t<graph_t::is_multi_gpu == false,
+                 std::tuple<device_vec_t<typename graph_t::vertex_type>,
+                            device_vec_t<typename graph_t::weight_type>,
+                            device_vec_t<index_t>>>
+random_walks_impl(raft::handle_t const& handle,
+                  graph_t const& graph,
+                  device_const_vector_view<typename graph_t::vertex_type, index_t>& d_v_start,
+                  index_t max_depth)
+{
+  using vertex_t = typename graph_t::vertex_type;
+  using edge_t   = typename graph_t::edge_type;
+  using weight_t = typename graph_t::weight_type;
+  using seed_t   = typename random_engine_t::seed_type;
+  using real_t   = typename random_engine_t::real_type;
+
+  vertex_t num_vertices = graph.get_number_of_vertices();
+  bool valid_start = thrust::all_of(rmm::exec_policy(handle.get_stream())->on(handle.get_stream()),
+                                    d_v_start.begin(),
+                                    d_v_start.end(),
+                                    [num_vertices] __device__(auto crt_vertex) {
+                                      return (crt_vertex >= 0) && (crt_vertex < num_vertices);
+                                    });
+
+  CUGRAPH_EXPECTS(valid_start == true, "Invalid set of starting vertices.");
+
+  auto nPaths = d_v_start.size();
+  auto stream = handle.get_stream();
+
+  random_walker_t<graph_t, random_engine_t> rand_walker{
+    handle, graph, static_cast<index_t>(nPaths), static_cast<index_t>(max_depth)};
+
+  // pre-allocate num_paths * max_depth;
+  //
+  auto coalesced_sz = nPaths * max_depth;
+  device_vec_t<vertex_t> d_coalesced_v(coalesced_sz, stream);  // coalesced vertex set
+  device_vec_t<weight_t> d_coalesced_w(coalesced_sz, stream);  // coalesced weight set
+  device_vec_t<index_t> d_paths_sz(nPaths, stream);            // paths sizes
+  device_vec_t<edge_t> d_crt_out_degs(nPaths, stream);  // out-degs for current set of vertices
+  device_vec_t<real_t> d_random(nPaths, stream);
+  device_vec_t<vertex_t> d_col_indx(nPaths, stream);
+  device_vec_t<vertex_t> d_next_v(nPaths, stream);
+  device_vec_t<weight_t> d_next_w(nPaths, stream);
+
+  // FIXME: abstract out seed initialization:
+  //
+  seed_t seed0 = static_cast<seed_t>(std::time(nullptr));
```

Comment:
Better to take seed as an input parameter for better reproducibility.

Reply:
Exposing this detail would propagate too high up into API level. Remember, the API that was agreed upon doesn't take a `seed` argument, and I think for good reasons (because that would push responsibility to the caller / user, who doesn't really care about that).

The way to use a certain (possibly fixed) seed (-ing mechanism) can be currently achieved by defining a custom `random_engine_t` that ignores the seeding mechanism currently in place and uses its own. Which was part of the reason for providing the `random_engine_t` policy.

Reply:
Fixed. I've added seed propagation in the `detail`. Seeding info is now returned for repro. (Note: As said, the API that was signed of on doesn't expose seeding).